File: Sprint4/C/C++/main.cpp
// https://contest.yandex.ru/contest/23991/problems/C/
#include <iostream>
#include <vector>
#include <string>

std::vector<long long> prefix_hashes;
std::vector<long long> powers;

void pre_calc_prefix_hashes(int a, int m, std::string& s) {
    int n = int(s.length());

    prefix_hashes.resize(n + 1);
    powers.resize(n + 1);

    powers[0] = 1;
    prefix_hashes[0] = 0;
    for (int i = 1; i <= n; i++) {
        powers[i] = (powers[i - 1] * a) % m;
        prefix_hashes[i] = (prefix_hashes[i - 1] * a + int(s[i - 1])) % m;
    }
}

long long substring_poly_hash(int m, int l, int r) {
    long long result;

    if (l == 1) {
        result = prefix_hashes[r];
    } else {
        result = (prefix_hashes[r] - prefix_hashes[l - 1] * powers[r - l + 1] + m) % m;
        if (result < 0) {
            if ( std::abs(result) % m != 0 ) {
                result += ((std::abs(result) / m) + 1) * m;
                result %= m;
            } else {
                result = 0;
            }
        }
    }

    return result;
}

int main() {
    int a, m;
    std::string s;

    std::cin >> a >> m;
    std::cin >> s;
    pre_calc_prefix_hashes(a, m, s);

    int t;
    std::cin >> t;
    for (int i = 0; i < t; i++) {
        int l, r;
        std::cin >> l >> r;
        std::cout << substring_poly_hash(m, l, r) << std::endl;
    }

    return 0;
}

File: Sprint4/C/Rust/src/main.rs
// https://contest.yandex.ru/contest/23991/problems/C/
use std::io::BufRead;

struct PrefixHashes {
    m: i64,
    hashes: Vec<i64>,
    powers: Vec<i64>,
}

impl PrefixHashes {
    fn new(a: i64, m: i64, s: String) -> Self {
        let mut hashes = Vec::with_capacity(s.len() + 1);
        let mut powers = Vec::with_capacity(s.len() + 1);

        powers.push(1i64);
        hashes.push(0i64);
        for i in 1..s.len() + 1 {
            powers.push((powers[i - 1] * a) % m);
            hashes.push((hashes[i - 1] * a + s.as_bytes()[i - 1] as i64) % m);
        }

        Self { m, hashes, powers }
    }

    fn substring_poly_hash(&self, l: usize, r: usize) -> i64 {
        if l == 1 {
            return self.hashes[r];
        }

        let mut result =
            ((self.hashes[r] - self.hashes[l - 1]) * self.powers[r - l + 1] + self.m) % self.m;

        if result < 0 {
            let abs_result = i64::abs(result);
            if abs_result % self.m != 0 {
                result += (abs_result / self.m + 1) * self.m;
                result %= self.m;
            } else {
                result = 0;
            }
        }

        result
    }
}

fn main() {
    let stdin = std::io::stdin();
    let mut lines = stdin.lock().lines();

    let a: i64 = lines.next().unwrap().unwrap().parse().unwrap();
    let m: i64 = lines.next().unwrap().unwrap().parse().unwrap();

    let s = lines.next().unwrap().unwrap();

    let prefix_hashes = PrefixHashes::new(a, m, s);

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    for _ in 0..t {
        let line = lines.next().unwrap().unwrap();
        let parts = line
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap())
            .collect::<Vec<usize>>();
        let l = parts[0];
        let r = parts[1];
        println!("{}", prefix_hashes.substring_poly_hash(l, r));
    }
}
